sessions/return-oriented-programming/README.md

```diff
@@ -0,0 +1,810 @@
+ 804a000 f6820408 06830408 16830408           ............
+```
+
+We recognize `f6820408` (`0x80482f6`) as being the next instruction in the
+`puts@plt` stub that we disassembled above. Which then pushes 0 in the stack and
+calls 0x80482e0. This is the call to the one-time resolver, and it looks like
+this:
+```
+$ objdump -D -j .plt -M intel hello | grep -A 3 '080482e0'
+
+080482e0 <puts@plt-0x10>:
+ 80482e0:	ff 35 f8 9f 04 08    	push   DWORD PTR ds:0x8049ff8
+ 80482e6:	ff 25 fc 9f 04 08    	jmp    DWORD PTR ds:0x8049ffc
+ 80482ec:	00 00                	add    BYTE PTR [eax],al
+```
+
+What's going on here? What's actually happening is lazy binding - by convention
+when the dynamic linker loads a library, it will put an identifier and
+resolution function into known places in the GOT. Therefore, what happens is
+roughly this: on the first call of a function, it falls through to call the
+default stub, it simply jumps to the next instruction. The identifier is pushed
+on the stack, the dynamic linker is called, which at that point has enough
+information to figure out “hey, this program is trying to find the function
+foo”. It will go ahead and find it, and then patch the address into the GOT such
+that the next time the original PLT entry is called, it will load the actual
+address of the function, rather than the lookup stub. Ingenious! 
+
+### Further Inspection
+Going further into the resolver is left as an exercise. You can use GDB to
+inspect the address in `0x8049ffc`, and what happens when this jumps there.
+
+## Return Oriented Programming (ROP)
+### Motivation
+In the previous sessions we discussed `ret2libc` attacks. The standard attack
+was to perform an overwrite in the following way:
+```
+RET + 0x00:   addr of system
+RET + 0x04:   JUNK
+RET + 0x08:   address to desired command (e.g. '/bin/sh')
+```
+
+However, what happens when you need to call multiple functions? Say you need
+to call `f1()` and then `f2(0xAB, 0xCD)`? The payload should be:
+```
+RET + 0x00:   addr of f1
+RET + 0x04:   addr of f2 (return address after f1 finishes)
+RET + 0x08:   JUNK (return address after f2 finishes: we don't care about what happens after the 2 functions are called)
+RET + 0x0c:   0xAB (param1 of f2)
+RET + 0x10:   0xCD (param2 of f2)
+```
+
+What about if we need to call `f1(0xAB, 0xCD)` and then `f2(0xEF, 0x42)`?
+```
+RET + 0x00:   addr of f1
+RET + 0x04:   addr of f2 (return address after f1 finishes)
+RET + 0x08:   0xAB (param1 of f1)
+RET + 0x0c:   0xCD (param2 of f1) but this should also be 0xEF (param1 of f2)
+RET + 0x10:   0x42 (param2 of f2)
+```
+
+### NOP Analogy
+While `ret2libc` uses functions directly, ROP uses a finer level of code
+execution: instruction groups. Let's explore an example:
+```c
+int main(void)
+{
+	char a[16];
+	read(0, a, 100);
+ 
+	return 0;
+}
+```
+
+This code obviously suffers from a stack buffer overflow. The offset to the
+return address is 28. So `DOWRD`s from offset 28 onwards will be popped from the

```

Comment:
Shouldn't `28` actually be `24`? You are running on `x32`, right? (the above stack representation is incremented in steps of 4, so I assumed that you want to build upon previous examples, therefore `x32`).